Add observable support with `symbol-observable`

mAAdhaTTah opened the pull request:
Handles returned observables the same way we handle returned promises.
Use the shared `symbol-observable` library for interop and consume the
stream until it ends. If an error is emitted, call callback with error object
from stream.

Fixes #1898.

mAAdhaTTah commented:
Bumping this. Would love to get Observable support in Mocha, similar to how Promises are handled. Anything I can do to help?

boneskull commented:
@mAAdhaTTah Thanks for the PR.  

Even though #1898 was created in 2015, observables are still in [stage 1](https://github.com/tc39/proposals).

Until the proposal reaches stage 3 (or 4 if we're cautious; see [process document](https://tc39.github.io/process-document/)), adding explicit support for them is, IMO, jumping the gun.

It's likely this will also break compatibility with older browsers that we still support.  It'd be a good reason (among many) *to* break compatibility with those browsers--when the feature is ready.

That's my take on this; if anyone else has another opinion, please speak up!  

cc @mochajs/core

* * *

As an aside, from an interface perspective, I'm interested in how best to handle observables which never complete.  It wouldn't make much sense to `return` one, since your test will timeout, though it's still valuable to make assertions against them.

Given that *some* observables could be returned and *others* could not, does it make sense to add support for this at all?  A `Promise` either fulfills or it doesn't.  Not so with observables.

mAAdhaTTah commented:
> Until the proposal reaches stage 3 (or 4 if we're cautious; see process document), adding explicit support for them is, IMO, jumping the gun.

Fair point. I thought I had read somewhere it was moving to stage 2 soon, but I don't recall. Even still, we're still a ways off from Observables getting specced.

> It's likely this will also break compatibility with older browsers that we still support.

It shouldn't; `symbol-observable` is a polyfill, so it checks if `Symbol` exists in the environment and falls back to a plain string should it not be found.

> A Promise either fulfills or it doesn't. Not so with observables.

Strictly speaking, you can write a Promise that never resolves:

```js
const myPromise = new Promise((resolve, reject) => {
    // do nothing
});

myPromise.then(() => {
    console.log('Will never get called');
});
```

But yes, you're much more likely to have an Observable that never ends than a Promise that never resolves. That said, one of the things I like about Observables is you can use something like `take` to only get the next X number of values out of it, which provides a kinda, sorta `plan` for your tests (although it won't help if the Observables accidentally produce extra values).

Other than that, I don't think you _can_ make assertions against an Observable that doesn't complete, or at least, I'm don't think Mocha can provide any special handling for it, similar to just timing out when a Promise fails to resolve.

Hope that addresses some of your concerns. If you'd prefer to wait until Observables reach stage 3+, I'll keep this PR open and look at providing this functionality through a library similar to how `mocha-as-promised` handled it.

mAAdhaTTah commented:
I'm going to close this. My approach to testing with observable is to use spies on `next` & `error` and verify the values in `complete`, where I call `done` manually. This provides better test planning than `take`, as extra values are allowed to flow through without ending the stream and ensures you're testing for the exact values you expect.

Also, most Observable implementations have a `toPromise` method, if you prefer to return the Observable, so there isn't really a good reason to add support to Mocha directly.

ORESoftware commented:
maybe someone could create `Observable.to(Promise)`, then return that in mocha hooks? Not sure if that's possible, but maybe.
